### ISSUE TITLE
Auth.currentAuthenticatedUser() does not return Cognito User.

### ISSUE THREAD

**Issue by irfanonk** [aws-amplify/amplify-js#11287]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication, REST API

### Amplify Categories

auth

### Environment information

<details>

```
# Put output below this line

System:
    OS: macOS 13.0.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 126.13 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.18.0 - ~/.nvm/versions/node/v16.18.0/bin/node
    npm: 8.19.2 - ~/.nvm/versions/node/v16.18.0/bin/npm
    Watchman: 2023.04.03.00 - /opt/homebrew/bin/watchman
  Browsers:
    Brave Browser: 111.1.49.132
    Chrome: 112.0.5615.137
    Edge: 112.0.1722.58
    Firefox: 111.0.1
    Safari: 16.1
  npmPackages:
    cors: ^2.8.5 => 2.8.5 
    express: ^4.18.2 => 4.18.2 
    google-auth-library: ^8.7.0 => 8.7.0 
    hardhat: ^2.12.2 => 2.12.2 
  npmGlobalPackages:
    @aws-amplify/cli: 11.0.3
    corepack: 0.14.1
    eas-cli: 2.8.0
    nodemon: 2.0.20
    npm: 8.19.2
    serve: 14.1.2
```

</details>

### Describe the bug

Login with google token using Auth.currentAuthenticatedUser() only returns  google token like so:

{
    "id": "us-east-1:xxxxx",
    "name": "xxx",
    "token": "xxxxxx"
}

### Expected behavior

it should return cognito user, or id,access and refresh tokens to create cognito user. 

### Reproduction steps

1. install aws-amplify
2. Login with google and get a valid google id token then pass it to Auth.federatedSignIn
3.  Get the current user with Auth.currentAuthenticatedUser

### Code Snippet

```javascript
// Put your code below this line.
    const googleToken = response.credential;
    const googleData = parseToken(googleToken);
    console.log('googleData:', googleData);
    const { email, given_name, exp } = googleData;
    Auth.federatedSignIn(
      'google',
      {
        token: googleToken,
        expires_at: exp
      },
      {
        name: given_name
      }
    ).then((response) => {
      console.log('federatedSignIn res', response);
      // Use the Auth.currentAuthenticatedUser() method to check if the user is already authenticated
      Auth.currentAuthenticatedUser()
        .then((cognitoUser) => {
          console.log('cognitoUser:', cognitoUser);
          const cognitoToken =
            cognitoUser.signInUserSession.accessToken.jwtToken;
          // Use the Cognito token to authenticate requests to your backend API
          console.log('cognitoToken:', cognitoToken);
        })
```

### Log output

<details>

```
// Put your logs below this line
{
    "id": "us-east-1:xxxxx",
    "name": "xxx",
    "token": "xxxxxx"
}

```

</details>

### Manual configuration

```
Amplify.configure({
  Auth: {
    // (required) only for Federated Authentication - Amazon Cognito Identity Pool ID
    identityPoolId: 'us-east-1:xxxx',

    // (required)- Amazon Cognito Region
    region: 'us-east-1',

    // (optional) - Amazon Cognito Federated Identity Pool Region
    // Required only if it's different from Amazon Cognito Region
    identityPoolRegion: 'us-east-1',

    // (optional) - Amazon Cognito User Pool ID
    userPoolId: 'us-east-1_xxx',

    // (optional) - Amazon Cognito Web Client ID (26-char alphanumeric string, App client secret needs to be disabled)
    userPoolWebClientId: 'xxx',

    // (optional) - Enforce user authentication prior to accessing AWS resources or not
    mandatorySignIn: false,

    // (optional) - Manually set key value pairs that can be passed to Cognito Lambda Triggers
    clientMetadata: { myCustomKey: 'myCustomValue' },

    // (optional) - Hosted UI configuration
    oauth: {
      domain: 'myapp.auth.us-east-1.amazoncognito.com',
      scope: [
        'phone',
        'email',
        'profile',
        'openid',
        'aws.cognito.signin.user.admin'
      ],
      redirectSignIn: `${window.location.origin}/login-google`,
      redirectSignOut: 'http://localhost:3000/',
      clientId: 'xxxxx',
      responseType: 'code' // or 'token', note that REFRESH token will only be generated when the responseType is code
    }
  }
});
```

### Additional information and screenshots

is there a way to create valid cognito user with Auth.federatedSignIn response. The response is like so :

```
{
    "identityId": "us-east-1:xx-xxxxx",
    "accessKeyId": "xxxx",
    "secretAccessKey": "xx/xxx/x+xxxx",
    "sessionToken":"xxxxxx"
}
```

**Comment by tannerabread** [comment]
Hi @irfanonk, because you are federating directly into the identity pool rather than the Cognito user pool, it will [not be possible to get a Cognito User](https://docs.amplify.aws/lib/auth/advanced/q/platform/js/#identity-pool-federation:~:text=the%20user%20you%20get%20after%20calling%20this%20method) returned from this method. 

What information were you looking for from the response? I was able to get some user information in an object with the following structure:
```js
email: "email@example.com"
id: "us-east-1:XXXXXXXXX"
name: "User Name"
token: "<token>"
```

Full app code:
```js
const SignInWithGoogle = () => {
  useEffect(() => {
    // Check for an existing Google client initialization
    if (!window.google && !window.google?.accounts) createScript();
  }, []);

  // Load the Google client
  const createScript = () => {
    const script = document.createElement("script");
    script.src = "https://accounts.google.com/gsi/client";
    script.async = true;
    script.defer = true;
    script.onload = initGsi;
    document.body.appendChild(script);
  };

  // Initialize Google client and render Google button
  const initGsi = () => {
    if (window.google && window.google?.accounts) {
      window.google.accounts.id.initialize({
        client_id: "<client_id>",
        callback: (response) => {
          getAWSCredentials(response.credential);
        },
      });
      window.google.accounts.id.renderButton(
        document.getElementById("googleSignInButton"),
        { theme: "outline", size: "large" }
      );
    }
  };

  // Exchange Google token for temporary AWS credentials
  const getAWSCredentials = async (credential) => {
    const token = jwt(credential);
    const user = {
      email: token.email,
      name: token.name,
    };
    await Auth.federatedSignIn(
      "google",
      { token: credential, expires_at: token.exp },
      user
    )
      .then((cred) => {
        // If success, you will get the AWS credentials
        console.log({ cred });
        return Auth.currentAuthenticatedUser();
      })
      .then((user) => {
        // If success, the user object you passed in Auth.federatedSignIn
        console.log({ user });
      })
      .catch((e) => {
        console.log(e);
      });
  };

  return (
    <div>
      <button id="googleSignInButton" />
    </div>
  );
};

```

**Comment by irfanonk** [comment]
i was expecting to create a valid Cognito User with response. I need to add or update user attributes. So is there a way to create Cognito User from this response? or I can get id_token and acces_token but still not able to do it.

**Comment by tannerabread** [comment]
It's possible to create the Cognito user, but you would need to use the `federatedSignIn` method like is [shown on this page](https://docs.amplify.aws/lib/auth/social/q/platform/js/)

So with google you would either do:
```js
// to open Hosted UI and log in through the UI
Auth.federatedSignIn();

// to bypass the Hosted UI 
Auth.federatedSignIn({ provider: 'Google' });
```

Doing it this way will create a user in your user pool rather than federating directly into the identity pool

**Comment by irfanonk** [comment]
This detail is important to know. Thank you. 
I should say still i am unable to achieve creating Cognito User with Auth.federatedSignIn({ provider: 'Google' }). 
When I use responseType: 'token', It returns only id_token and access_token, there is no refresh token so i could not create a valid Cognito User and cognito session. when it comes to responseType: 'code', i am getting invalid_request error from token endpoit for grant_type :authorization_code. this is the config and codes for exchanging code with token. 

```
const AUTH_DOMAIN = 'https://xxx.auth.us-east-1.amazoncognito.com';

    const grantType = 'authorization_code';
    const clientId = 'xxx'; 
 const clientSecret = 'xxxx',
    const redirectUri = `${window.location.origin}/login-google`; 
    
    axios
      .post(
        `${AUTH_DOMAIN}/oauth2/token`,

        new URLSearchParams({
          grant_type: grantType,
          code: code,
          state: state,
          client_id: clientId,
          redirect_uri: redirectUri
        }),
        {
          headers: {
            Authorization: getBase64EncodedCredential(clientId, clientSecret)
          }
        }
      )
      .then((response) => {
        // handle success
        console.log(response.data);
      })
      .catch((error) => {
        // handle error
        console.error(error);
      });
      
      function getBase64EncodedCredential(cognitoAppId, cognitoAppSecret) {
    return 'Basic ' + btoaImplementation(cognitoAppId + ':' + cognitoAppSecret);
  }
  function btoaImplementation(str) {
    try {
      return btoa(str);
    } catch (err) {
      return Buffer.from(str).toString('base64'); 
    }
  }
```
I do not understand where i am doing wrong.
Any help for this would be much appreciated.

**Comment by nadetastic** [comment]
Hi @irfanonk - its interesting that `Auth.federatedSignIn({ provider: 'Google'})` did not create a cognito user pool user for you. Did you see any errors or response from this?

Also I want to point out the distinction it the two methods discussed so far.

1. Authenticate directly with Google and retrieve tokens from Google to pass to Cognito - This doesn't auto create a cognito user for you and is the method you were initially using with:
```js
Auth.federatedSignIn(
	'google',
	{ token: '<token_from_google>', expires_at: exp },
	{ name: given_name }
)
```
2. Authenticate through Cognito but never get tokens from Google - This will create a Cognito user for you with the name `google_xxxx...` and you will not need to handle any tokens from Google, Cognito will do that for you by doing the following:
```js
Auth.federatedSignIn({ provider: 'Google' })
```

**Comment by irfanonk** [comment]
Hi @nadetastic 
No, I am not saying it's not creating Cognito User in user pool. It returns id_token and access_token which indicates its creating a new user in user pool. 

What is problem here is that I cannot create a new Cognito User instance in front app.

As far as I understand, since i need to update user attributes so I have to create a valid cognito user and cognito session in front. So to be able initiate new cognito session in front app I need to id_token, access_token and refresh_token. Above approach that is exchange code with token using token endpoint always returns invalid_request. 

1. First method returns session token like so:
`{
    "identityId": "us-east-1:xx-xxxxx",
    "accessKeyId": "xxxx",
    "secretAccessKey": "xx/xxx/x+xxxx",
    "sessionToken":"xxxxxx"
}`

But still it seems it is not possible to create a new cognito session with this data

2. Second method  
   a. returns id_token, and access_token when configured  responseType="token". I cannot create a valid session with these tokens
   b. returns code  responseType="code". Docs states that only way to get refresh token along with other tokens is that using grant_type="authorization_code".   I am trying to exchange authorization code with token but always get invalid_request. I have asked a lot of place but still stuck here. It is getting boring :)

ps: It might be related about this issue. [#8632](https://github.com/aws-amplify/amplify-js/issues/8632)

**Comment by irfanonk** [comment]
Is there any progress or suggestion about this problem?

**Comment by edgardcooper-lukasdemeyere** [comment]
We have the same problem. Sometimes we are able to log in, sometimes not. Looks like it's 50/50 whether the federated login works, when we authenticate and we are redirected, Amplify make 4 calls to https://cognito-domain/oauth2/token (eventhough I would expect only 1). When you're lucky 1 call is resolved successfully. The rest of the calls fail, either with invalid_grant or invalid_request, seems 50/50 which error code we get back.

The signout redirect fails as we have redirects in our project middleware. But the localStorage Cognito values are cleared, so suspect this isn't a problem. 

A video showing this behaviour:
https://youtu.be/xCLVCwc1M1M

**Comment by nadetastic** [comment]
@irfanonk the first method will not create a Cognito user - i can confirm that. The second method will create a cognito user for you however. So Im a bit confused as to where the issue is.

It sounds like you are trying to use the second method to create a Cognito user from the authorization code that is provided in the redirect URL after federating? If so you shouldn't need to create one as one is already available as long as the call is successful. You can use `Auth.currentAuthenticatedUser()` to see the cognito user.

Please clarify how you are encountering the `invalid_request` error message 🙂

**Comment by irfanonk** [comment]
Sure.  I am folowing Authorization code grant flow.

1. step: 
 call Auth.federatedSignIn with  responseType: 'code'. It returns a "code" in url.
2. step: 
send a new request to oauth2/token endpoint with grant_type = 'authorization_code'. It is supposed to return id_token_ access_token and refresh_token. 

Below is detailed explanation:

`
Amplify.configure({
  Auth: {
    identityPoolId: ENV.COGNITO_IDENTITY_ID,

    region: ENV.COGNITO_REGION,
    identityPoolRegion: ENV.COGNITO_REGION,
    userPoolId: ENV.COGNITO_USER_POOL_ID,

    userPoolWebClientId: 'xxxxx', 
    mandatorySignIn: false,
    oauth: {
      domain: ENV.COGNITO_DOMAIN,
      scope: [
        'phone',
        'email',
        'profile',
        'openid',
        'aws.cognito.signin.user.admin'
      ],
      redirectSignIn: `${window.location.origin}/login-google`,
      redirectSignOut: 'http://localhost:3000/',
      clientId: 'xxxxx', 
      responseType: 'code' // or 'token', note that REFRESH token will only be generated when the responseType is code
    }
  }
});
`
` await Auth.federatedSignIn({
        provider: CognitoHostedUIIdentityProvider.Google
      });
      `
    
  when login with above method (Auth.federatedSignIn), it creates a new user in user pool and returns id_token and access_token when responseType is 'token'.  However i need refresh token too. otherwise i can not create a new CognitoUser object at client side ( so not able to update user attributes as well since not possible to create a valid CognitoUser with id_token and access_token only). Hence i need that REFRESH TOKEN too. According to  docs, [for example this one ](https://aws.amazon.com/blogs/mobile/understanding-amazon-cognito-user-pool-oauth-2-0-grants/) in order to get refresh token after federated sign in once should configure responseType as this : responseType: 'code'. 
   
  when you configure responseType: 'code' you will get  "code" and "state" variables in the url in return. The docs says that it is possible  to get  id_token, access_token and refresh_token all together by using this "code" with sending a request to /oauth2/token endpoint. It also emphasizes that if the client app that was used requires a secret, the Authorization header for this request is set as “Basic BASE64(CLIENT_ID:CLIENT_SECRET)“, where BASE64(CLIENT_ID:CLIENT_SECRET) is the base64 representation of the app client ID and app client secret, concatenated with a colon.   whatever i did, i failed to get those tokens since i got always invalid_request. If I send a new request with already used "code", this time it returns "invalid_grant". The sample request is here like so: 
  
  `
  const AUTH_DOMAIN = 'https://xxx.auth.us-east-1.amazoncognito.com';

    const grantType = 'authorization_code';
    const clientId = 'xxx'; 
 const clientSecret = 'xxxx',
    const redirectUri = `${window.location.origin}/login-google`; 
    
    axios
      .post(
        `${AUTH_DOMAIN}/oauth2/token`,

        new URLSearchParams({
          grant_type: grantType,
          code: code,
          state: state,
          client_id: clientId,
          redirect_uri: redirectUri
        }),
        {
          headers: {
            Authorization: getBase64EncodedCredential(clientId, clientSecret)
          }
        }
      )
      .then((response) => {
        // handle success
        console.log(response.data);
      })
      .catch((error) => {
        // handle error
        console.error(error);
      });
      
      function getBase64EncodedCredential(cognitoAppId, cognitoAppSecret) {
    return 'Basic ' + btoaImplementation(cognitoAppId + ':' + cognitoAppSecret);
  }
  function btoaImplementation(str) {
    try {
      return btoa(str);
    } catch (err) {
      return Buffer.from(str).toString('base64'); 
    }
  }
  `
  
  Auth.currentAuthenticatedUser() returns only token of provider (here google token). So it is not useful fro me. it is something like this :
  
  `
{
    "id": "us-east-1:xxxxx",
    "name": "xxx",
    "token": "xxxxxx"
}
  `
  
  
    **In short, I am seeking to get that refresh token along with id token and access token from cognito user pool.**
   
   Jeremy Glesner addressed this problem here in his [article](https://medium.com/codex/how-to-process-an-aws-cognito-authorization-code-grant-using-aws-amplify-b49d9ee052ca) and here in his [issue](https://github.com/aws-amplify/amplify-js/issues/8632#issuecomment-886133768)
   
   another issue opened [here](https://github.com/aws-amplify/amplify-js/issues/6347)

**Comment by irfanonk** [comment]
any help would be much appreciated

**Comment by nadetastic** [comment]
@irfanonk You should be able to retrieve all the session token you need by calling `Auth.currentSession()` have you had a chance to try this? It returns the following object that contains the refresh token:

```json
{
	"accessToken": {},
	"clockDrift": "<number-value>"
	"idToken": {}
	"refreshToken" : {
		"token": "<jwt-token>"
	}
}
```

**Comment by irfanonk** [comment]
Thanks for your response.
I have tried all possible methods something like this:

the method I call for google sign in:

```js
  const handleGoogleLoginSuccess = async (response) => {
    console.log('response:', response);
    const googleToken = response.credential;
    const googleData = parseToken(googleToken);
    console.log('googleData:', googleData);
    const { email, given_name, exp } = googleData;

    try {
      const cognitoResponse = await Auth.federatedSignIn(
        'google',
        { token: googleToken, expires_at: exp },
        {
          name: given_name
        }
      );
      console.log('cognitoResponse:', cognitoResponse);
    } catch (error) {
      console.log('error:', error);
    }
  };
```
the response I go from methods:

```js
  useEffect(() => {
    const unsubscribe = Hub.listen('auth', ({ payload: { event, data } }) => {
      switch (event) {
        case 'signIn':
          console.log('signIn:', data);
          break;
        case 'signOut':
          console.log('signOut:', data);
          break;
        case 'customOAuthState':
          console.log('customOAuthState:', data);
          break;
        case 'tokenRefresh':
          console.log('tokenRefresh:', data);
          break;
      }
    });

    Auth.currentAuthenticatedUser()
      .then((currentUser) => {
        console.log('currentUser', currentUser);
        //   {
        //     "id": "us-east-1:xxx-xxx-xx-xx-xxxx",
        //     "name": "Ä°rfan",
        //     "token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjJkOWE1ZWY1YjEy......" google token
        // }
      })
      .catch((err) => console.log('No currentAuthenticatedUser', err));

    Auth.currentSession()
      .then((session) => {
        console.log('currentSession', session);
      })
      .catch((err) => console.log('No currentSession', err)); // No currentSession No current user

    Auth.currentCredentials()
      .then((creds) => {
        console.log('creds', creds);
        //   {
        //      "id": "us-east-1:xxx-xxx-xx-xx-xxxx",
        //     "accessKeyId": "xxxxx",
        //     "secretAccessKey": "/xxxxxx+",
        //     "sessionToken": "IQoJb3JpZ2luX2VjEFcaCXVzLWVhc3QtMSJGMEQCIGdf5gYxDJgwDnby.........",
        //     "expiration": "2023-05-19T23:07:13.000Z",
        //     "authenticated": true
        // }
      })
      .catch((err) => console.log('No creds', err));

    Auth.currentUserCredentials()
      .then((userCreds) => {
        console.log('userCreds', userCreds);
        //   {
        //      "id": "us-east-1:xxx-xxx-xx-xx-xxxx",
        //     "accessKeyId": "xxxxx",
        //     "secretAccessKey": "/xxxxxx+",
        //     "sessionToken": "IQoJb3JpZ2luX2VjEFcaCXVzLWVhc3QtMSJGMEQCIGdf5gYxDJgwDnby.........",
        //     "expiration": "2023-05-19T23:07:13.000Z",
        //     "authenticated": true
        // }
      })
      .catch((err) => console.log('No userCreds', err));

    Auth.currentUserInfo()
      .then((userInfo) => {
        console.log('userInfo', userInfo); //userInfo null
      })
      .catch((err) => console.log('No userInfo', err));

    Auth.currentUserPoolUser()
      .then((poolUser) => {
        console.log('poolUser', poolUser);
      })
      .catch((err) => console.log('No poolUser', err)); // No poolUser No current user

    return unsubscribe;
  }, []);
```

when I call other way of federated sign in like so:

```js
            Auth.federatedSignIn({
              provider: CognitoHostedUIIdentityProvider.Google
            });
```

any methods return anything but fall in catch block.

Is there a way to  use sessionToken and secretAccessKey?

**Comment by nadetastic** [comment]
@irfanonk I'm curious how your application is setup - would it be possible to share a sample of you app repo with me so I can take a look?

**Comment by irfanonk** [comment]
Unfortunately I am not able to share repo. yet I have already shared how I configured, logged in with google and tried to get refresh token. 

If possible share me a working example of **### authorization code flow** when login with google token (or any other social login). I need to see login with google where it returns authorization code that will be later exchanged with cognito tokens (id, access and refresh token).

if you are able to share that working example I could check it and try to understand where i am doing wrong for authorization code flow and share the result here.

**Comment by nadetastic** [comment]
@irfanonk ive setup this [repo branch](https://github.com/nadetastic/vite-amplified-react/blob/11287/src/App.jsx) where you can see how what we have discussed so far works in a React app.

**Comment by gandonov** [comment]
Having same issue here. We have a Cordova App with google social plugin that returns idToken; taking user to the webview or external browser as opposed to native android/ios login dialog would not be feasible in our case. 

We hold some user data in dynamodb that we would like to retrieve after authenticating user's email with google access token / tokenId . Would you know if this is possible to achieve in Amplify without invoking 

Auth.federatedSignIn({
    provider : "Google"
});

Any help is greatly appreciated

**Comment by gandonov** [comment]
**Update**: I was able to leverage Auth with aforementioned restrictions, by creating following flow:

on the client:
1. user logs into google, get id_token
2. send id_token over to the backend

on the backend:
3. verify id_token on the backend (for node.js I've used 'google-auth-library' npm package)
4. create a cognito user and record real email in some custom attribute, confirm user, set random password
5. send username/ random password to user

on the client:
6. Auth.signIn({ username, password} )

One caveat - 
I've noticed if we were to use Auth.federatedSignIn({ provider : "Google" })   cognito users are created with username
google_192301930123 where email attribute could be actual email. When I try to create that, it tells me that username must match look like email. How are users such as google_192301930123 generated? Am I missing something trivial and the above flow is overkill?

Thanks in advance for any input on this matter

**Comment by irfanonk** [comment]
hi @gandonov.

When you follow this:
```js
Auth.federatedSignIn(
        'google',
        { token: googleToken, expires_at: exp },
        {
          name: given_name
        }
      );
```
what you have done could be possible to create a new congito user in the user pool, otherwise congito do not create a new user in the user pool but in the identity pool.  It could be overkill because  Auth.federatedSignIn({ provider : "Google" })  already handle for you what you have done. Besides it seems you are sending password to client which is a security concern. 

When you configure responseType="token", Cognito creates a new user in user pool and returns Conginto idToken and accessToken which I think enough for your app.  So what you have noticed is OK.

But in my case, i need to refreshToken which according to docs could be obtained only by following "authorization code flow" that is only be carried out by setting responseType="code" in the config. ie Auth.configure({...,responseType="code"}).

**Comment by nadetastic** [comment]
With the release of the latest major version of Amplify (aws-amplify@>6), this should no longer be an issue! Please refer to our [release announcement](https://aws.amazon.com/blogs/mobile/amplify-javascript-v6/?sc_channel=sm&sc_publisher=TWITTER&sc_geo=GLOBAL&sc_outcome=awareness&linkId=248500593), [migration guide](https://docs.amplify.aws/react/build-a-backend/auth/auth-migration-guide/#authverifytotptoken), and [documentation ](https://docs.amplify.aws/javascript/start/getting-started/)for more information.
